scipy/stats/_quantile.py

```diff
@@ -257,3 +295,43 @@ def quantile(x, p, *, method='linear', axis=0, nan_policy='propagate', keepdims=
         res = xp.squeeze(res, axis=axis)
 
     return res[()] if res.ndim == 0 else res
+
+
+def _quantile_hf(y, p, n, method, xp):
+    ms = dict(inverted_cdf=0, averaged_inverted_cdf=0, closest_observation=-0.5,
+              interpolated_inverted_cdf=0, hazen=0.5, weibull=p, linear=1 - p,
+              median_unbiased=p / 3 + 1 / 3, normal_unbiased=p / 4 + 3 / 8)
+    m = ms[method]
+    jg = p*n + m - 1
+    j = jg // 1
+    g = jg % 1
+    if method == 'inverted_cdf':
+        g = xp.astype((g > 0), jg.dtype)
+    elif method == 'averaged_inverted_cdf':
+        g = (1 + xp.astype((g > 0), jg.dtype)) / 2
+    elif method == 'closest_observation':
+        g = (1 - xp.astype((g == 0) & (j % 2 == 1), jg.dtype))
+    if method in {'inverted_cdf', 'averaged_inverted_cdf', 'closest_observation'}:
+        g = xp.asarray(g)
+        g[jg < 0] = 0
+    j = xp.clip(j, 0., n - 1)
+    jp1 = xp.clip(j + 1, 0., n - 1)
+    j = xp.astype(j, xp.int64)
+    jp1 = xp.astype(jp1, xp.int64)
```

Comment:
why is this needed?

In general, it's not very easy to follow which variables are integer and which are floating-point. Consider adding some code comments to this effect.

Reply:
All variables are floating point until this line. (Even `n` because the array API standard doesn't support operations between float and integer arrays. And yeah `n` is an array in general because we have different numbers of non-NaN, non-masked elements in each slice.)

Right below, we use these in `take_along_axis`, so we have to ensure integer dtype. I could try to put these conversions into those lines, if that would help? Then there's no question that everything is float throughout.